Adding help section for chat

AugierLe42e opened the pull request:
Problem with linking :/
![final2](https://cloud.githubusercontent.com/assets/3429231/6196922/3cc65e1c-b3de-11e4-9b85-662a6aad8db9.png)

jhass commented:
The help section is clearly user oriented, we shouldn't include items for the podmin, in fact we shouldn't even display this section if the chat is disabled.

christophehenry committed:
Adding help section for chat

AugierLe42e commented:
Does client side has access to the configuration of the pod ?

jhass commented:
Not directly, that would be dangerous. I'm sure @zauberstuhl knows a way to detect it from the client side though.

Zauberstuhl commented:
The first thing which pops into my mind is that you could check whether `jsxc` variable exists.

AugierLe42e commented:
If not, I can modify the help controller to pass whether tha chat is active or not in a data-section, for instance ?

jhass commented:
Yes, that would be the alternative if there's no general way of detecting it from the client side. You would use gon for that.

AugierLe42e commented:
How about the linking problem ? Seems like the HTML is escaped :/

svbergerem commented:
@AugierLe42e Your images include English text. We decided some time ago that we don't want to add images to the help section which would need a translation.

AugierLe42e commented:
@svbergerem : That's right. Let's see if I can include it in HTML.

Zauberstuhl commented:
Checked it and it is not possible. You should go with the controller

AugierLe42e commented:
Thank you.

AugierLe42e commented:
Works but still having troubles with Jasmine tests and link injections.

jhass commented on app/controllers/help_controller.rb:
```diff
@@ -1,4 +1,10 @@
+  layout -> (c) { request.format == :mobile ? "application" : "with_header_with_footer" }
+
+  def faq
+    @data_chat = AppConfig.chat.enabled ? "enabled" : "disabled"
```
Let's use [gon](https://github.com/gazay/gon) here.

AugierLe42e replied:
Is it a usual use-case in d* ?
It seems a bit over-engineered to use a another just for that, right ?

jhass replied:
It's already available and solves the problem better than your current solution. I don't see why that would be over-engineered, in fact it's less code you have to add in this PR.

AugierLe42e replied:
> in fact it's less code you have to add in this PR.

In fact it's more code I have to remove. But if you think it's better, then let's do this.

jhass replied:
The changeset counts, not your work towards it ;P

jhass commented:
Could also use a rebase.

AugierLe42e commented on app/assets/javascripts/app/views/help_view.js:
```diff
@@ -187,8 +197,23 @@ app.views.Help = app.views.StaticContentView.extend({
+  },
+
+  chatEnabled: function(){
+    return gon.chat_enabled === 'enabled';
```
This doesn't work at all.

jhass commented on app/controllers/help_controller.rb:
```diff
@@ -1,4 +1,8 @@
+  layout -> (c) { request.format == :mobile ? "application" : "with_header_with_footer" }
+
+  def faq
+    gon.preloads[:chat_enabled] = AppConfig.chat.enabled ? 'enabled' : 'disabled'
```
Here you use the key `gon.preloads[:chat_enabled]` but in the JS you use the key `gon.chat_enabled`. I find the key `gon.chat_enabled` better, so chance this instance.

jhass replied:
Also passing `true`/`false` to to it as returned by `AppConfig.chat.enabled?` should be just fine.

AugierLe42e commented on app/controllers/help_controller.rb:
```diff
@@ -1,4 +1,8 @@
+  layout -> (c) { request.format == :mobile ? "application" : "with_header_with_footer" }
+
+  def faq
+    gon.chat_enabled = AppConfig.chat.enabled?
```
Doesn't work either. JS variable is `undefined` which make me think the variable is not corretly passed to the JS script.

jhass replied:
Looks like gon converts that to headlessCamlCase.

![](http://cloud.aeshna.de/u/mrzyx/screenshots/screenshot_20150222_000933.png)

AugierLe42e replied:
Weird... Thanks for the tip. Works no. What about the tags not being injected correctly ?

![injection](https://cloud.githubusercontent.com/assets/3429231/6316520/46232b88-ba29-11e4-829b-a4c11e8e0766.png)

jhass replied:
Handlebars feature. `{{}}` escapes, `{{{}}}` does not.

AugierLe42e commented on spec/javascripts/app/views/help_view_spec.js:
```diff
@@ -134,4 +141,28 @@ describe("app.views.Help", function(){
     });
   });
+
+describe("chat section", function(){
```
You are right, milou !

AugierLe42e commented:
Two last things and we are ok. [Images are not correctly included](https://github.com/diaspora/diaspora/pull/5665/files#diff-7425515578ef82f0d52e7bf0b0bacbf9R211) and [this Jasmin test still fails](https://github.com/diaspora/diaspora/pull/5665/files#diff-41cc85e93f2018f87904ee1daccc7280R146)

![jasmin](https://cloud.githubusercontent.com/assets/3429231/6316571/4cf36dd6-ba2b-11e4-84c0-58fdcfb0a966.jpg)

jhass commented on app/assets/javascripts/app/views/help_view.js:
```diff
@@ -187,8 +197,23 @@ app.views.Help = app.views.StaticContentView.extend({
+  },
+
+  linkImage: function(src, alt){
+    return '<div class="help-image"><img src="images/' + src + '" alt="' + alt + '"/></div>';
```
We got [js_image_paths](https://github.com/jhass/js_image_paths#usage), use it here.

AugierLe42e replied:
Is there something special to do for an image to be taken in account ?
Because :
![injection1](https://cloud.githubusercontent.com/assets/3429231/6316627/fc34c194-ba2d-11e4-87fb-1bcc909b49c7.png)
and though :
![injection2](https://cloud.githubusercontent.com/assets/3429231/6316628/03bdc24e-ba2e-11e4-8032-bae0c42cacc6.png)

jhass replied:
Does `public/assets/` exist? Try removing it. Also try `bin/rake tmp:cache:clear` and restarting the app server.

AugierLe42e replied:
Yay !
![help](https://cloud.githubusercontent.com/assets/3429231/6316650/359c5388-ba2f-11e4-8bac-dc28125f786c.png)

svbergerem replied:
In most cases (when there are less than 20 contacts in the aspect) there will be also the mail icon in the control bar. The really important thing is the chat icon. I am still not convinced that we should add an image here. Why don't we describe where you will find the button (Go to the contacts page, select an aspect and click on the chat icon in the icon bar under the header to enable chat for the aspect) and add the original entypo icon?

AugierLe42e commented:
@svbergerem : Why do you want to make me do complicated things where an image is sufficient ? I followed your advice : I dropped the images with english text and I added your description. I don't think this is that important, really.

svbergerem commented:
@AugierLe42e You added a screenshot with 3 icons. As I already said in most cases there will also another icon (mail) next to those. We might add/remove/change icons in the future and then we will have to change the screenshot. The 3 icons from the screenshot are already in the browser's cache.

Instead of loading them another time as an image I suggested to add them as an icon. I think it would be sufficient if we would only include the chat icon but we could also display all 3 or 4 icons. For the code it doesn't really make a difference.

Instead of adding an image 
`<div class="help-image"><img src="' + ImagePaths.get(src) + '" alt="' + alt + '"/></div>`
you would have to add an icon
`<i class="entypo chat"></i>`.
Why is that more complicated?

Adding those icons via entypo icons instead of an image makes it also easier to change the appearance of the icon later. We want to make the icon red? Just add `color: red`. We want to change the icon size? Just change `font-size`. If we would like to do the same with the screenshot we would either have to scale or edit the image. Both won't be as efficient.

AugierLe42e commented:
You may be right.

AugierLe42e commented:
@svbergerem : done. Will you help me to solve jASMIN TEST

AugierLe42e commented on app/assets/javascripts/app/views/help_view.js:
```diff
@@ -187,8 +197,31 @@ app.views.Help = app.views.StaticContentView.extend({
+              <i class="entypo lock-open"></i> \
+              <i class="entypo chat"></i> \
+              <i class="entypo trash"></i> \
+            </div>'
```
Oh dear...

svbergerem commented on spec/javascripts/app/views/help_view_spec.js:
```diff
@@ -134,4 +141,28 @@ describe("app.views.Help", function(){
+      });
+
+      it('should display the chat', function(){
+        expect(this.view.$el.find('div[data-section=chat]').length).toBe(1);
```
use `a[data-section=chat]` instead.

svbergerem commented on spec/javascripts/app/views/help_view_spec.js:
```diff
@@ -134,4 +141,28 @@ describe("app.views.Help", function(){
+      });
+
+      it('should not display the chat', function () {
+        expect(this.view.$el.find('div[data-section=chat]').length).toBe(0);
```
use `a[data-section=chat]` instead.

svbergerem commented on spec/javascripts/app/views/help_view_spec.js:
```diff
@@ -134,4 +141,28 @@ describe("app.views.Help", function(){
+  describe("chat section", function(){
+    describe("chat enabled", function(){
+      beforeEach(function(){
+        gon.chatEnabled = true;
```
In the help view you set `chat_enabled` on initialization. Because of that it doesn't change when you change `gon.chatEnabled` afterwards.

Add `this.view = new app.views.Help();` after this line to fix that.

AugierLe42e replied:
Yes it changes because `gon` changes snake_case to headlessCamelCase. This is awkward but meh... I should change the help view instead.

svbergerem replied:
I don't think you should. There is no need to change `gon.chatEnabled` on a normal setup after loading the page so `this.chat_enabled` doesn't have to change either.

`this.chat_enabled` doesn't change after loading the page because you copy the value of `gon.chatEnabled` on initialization and don't change `this.chat_enabled` afterwards. (which is fine as I already wrote) You just have to make sure that in this particular case `this.chat_enabled` changes which you can do by adding `this.view = new app.views.Help();` here.

AugierLe42e replied:
Ok. I thought chat `this.chat_enabled` was just an alias to the function `chatEnabled()` and was called each time the page is rendered.

svbergerem commented on spec/javascripts/app/views/help_view_spec.js:
```diff
@@ -134,4 +141,28 @@ describe("app.views.Help", function(){
+
+    describe("chat enabled", function(){
+      beforeEach(function(){
+        gon.chatEnabled = false;
```
Add `this.view = new app.views.Help();` after this line.

AugierLe42e commented on spec/javascripts/app/views/help_view_spec.js:
```diff
@@ -134,4 +141,30 @@ describe("app.views.Help", function(){
+      beforeEach(function(){
+        this.chat_enabled = false;
+        this.view = new app.views.Help();
+        this.view.render();
```
@svbergerem : Unless I've missed something, [this does not work](https://travis-ci.org/diaspora/diaspora/jobs/51999707#L570)

AugierLe42e commented:
\o/ @svbergerem : thanks, works now !
@jhass : you can merge this.

svbergerem commented on app/assets/javascripts/app/views/help_view.js:
```diff
@@ -187,8 +197,31 @@ app.views.Help = app.views.StaticContentView.extend({
+
+  linkImage: function(src, alt){
+    return '<div class="help-image"><img src="' + ImagePaths.get(src) + '" alt="' + alt + '"/></div>';
+  },
```
We don't use that function anymore. Should we still add it?

AugierLe42e replied:
It may be useful later. Now it is written, why should we remove it ?

jhass replied:
Because with that argument we would still maintain about 6-10kloc more code in this repository. Please remove it.

svbergerem commented on app/assets/javascripts/app/views/help_view.js:
```diff
@@ -187,8 +197,31 @@ app.views.Help = app.views.StaticContentView.extend({
+              <i class="entypo lock-open"></i> \
+              <i class="entypo chat"></i> \
+              <i class="entypo trash"></i> \
+            </div>';
```
I agree with Hound here. I'd prefer
```js
return '<div class="help-chat-icons">' +
       '  <i class="entypo lock-open"></i>' +
       '  <i class="entypo chat"></i>' +
       '  <i class="entypo trash"></i>' +
       '</div>';
```

svbergerem commented on app/views/help/faq.html.haml:
```diff
@@ -3,9 +3,9 @@
   = t('_help')
 
-#help
+#help
```
These are no important changes in this file so please don't commit them.

svbergerem commented on config/locales/diaspora/en.yml:
```diff
@@ -449,6 +450,10 @@ en:
+    chat:
+      title: "Chat"
+      add_contact_roster_q: "How do I add contacts to my roster ?"
+      add_contact_roster_a: "Go to the %{contacts}, select an aspect and click on the chat icon to enable chat for the aspect. %{toggle_privilege} Or when you create a new aspect we will also ask you if you would like to enable chat for that aspect"
```
Maybe 'Go to the contacts page, ...'? In that case I'd remove `help.contacts` and add `help.chat.contacts_page` instead.

svbergerem commented on spec/javascripts/app/views/help_view_spec.js:
```diff
@@ -8,6 +8,13 @@ describe("app.views.Help", function(){
     Diaspora.Page = "HelpFaq";
   });
 
+  describe("Fonctions", function(){
```
typo

AugierLe42e commented:
Ok now ?

svbergerem commented on config/locales/diaspora/en.yml:
```diff
@@ -449,6 +449,11 @@ en:
+      title: "Chat"
+      contacts_page: "contacts"
+      add_contact_roster_q: "How do I add contacts to my roster ?"
+      add_contact_roster_a: "Go to the %{contacts_page} page, select an aspect and click on the chat icon to enable chat for the aspect. %{toggle_privilege} Or when you create a new aspect we will also ask you if you would like to enable chat for that aspect"
```
I think this should be 'Go to the %{contacts_page}, select ...' and `contacts_page` should be 'contacts page'. Otherwise there might be languages where the current solution doesn't work.

svbergerem commented:
Apart from the comment I just added this looks good. I have no time right now to test this PR and merge it afterwards so either you will have to wait a few days or someone else merges this PR after you fixed the final remark.

jhass commented on config/locales/diaspora/en.yml:
```diff
@@ -449,6 +449,11 @@ en:
+    chat:
+      title: "Chat"
+      contacts_page: "contacts page"
+      add_contact_roster_q: "How do I add contacts to my roster ?"
```
No [Plenken](http://en.wikipedia.org/wiki/Plenken) please.

jhass commented on config/locales/diaspora/en.yml:
```diff
@@ -449,6 +449,11 @@ en:
+      title: "Chat"
+      contacts_page: "contacts page"
+      add_contact_roster_q: "How do I add contacts to my roster ?"
+      add_contact_roster_a: "Go to the %{contacts_page}, select an aspect and click on the chat icon to enable chat for the aspect. %{toggle_privilege} Or when you create a new aspect we will also ask you if you would like to enable chat for that aspect"
```
Also missing period at the end. How about "for this aspect", or "make contacts of this aspect in the chat available"? I'd drop the "Or " from the next sentence and say "you can choose to" instead of "we will ask you".

AugierLe42e commented:
Corrected.

jhass commented on config/locales/diaspora/en.yml:
```diff
@@ -449,6 +449,11 @@ en:
+      title: "Chat"
+      contacts_page: "contacts page"
+      add_contact_roster_q: "How do I add contacts to my roster?"
+      add_contact_roster_a: "Go to the %{contacts_page}, select an aspect and click on the chat icon to enable chat for the aspect. %{toggle_privilege} When you create a new aspect you can choose to enable chat for that aspect"
```
Still missing period for the second sentence. Looking at it again, the second "the aspect" in the first sentence seems a bit redundant maybe just replace it with "it". Also maybe the second sentence might better end in something like "for it too."

AugierLe42e replied:
Maybe :

*When you create a new one you can choose to enable chat for it* ?

jhass replied:
Guess we need @goobertron's help here ;)

goobertron replied:
Hi! I'm here.

Is it just the one question and answer being added, or is there more text elsewhere? This one seems to cover only how to enable chat for an entire aspect, not how to actually add an individual contact to your chat roster. Will there be, or should there be, help available for other aspects of the chat feature?

Is the list of people in your chat interface officially called a 'roster', or is that your term, Augier? If the term doesn't appear in the chat interface, I'll probably choose another one as 'roster' isn't the most common term.

AugierLe42e replied:
> Will there be, or should there be, help available for other aspects of the chat feature?

Theorically, yes. The first that come to my mind is how to access the chat feature from an external client. And then, maybe, how to use video/audio.

> Is the list of people in your chat interface officially called a 'roster', or is that your term, Augier?

Nope, this is the official JSXC terminology and this the term that appears on the wiki. I may not be the best term, though.

The question may not be the right one, actually. What about : "*How do I fill my friends list ?*"

goobertron replied:
How about something like:

> ### How do I chat with someone in diaspora*?
> 
> - First you need to enable chat for one of the aspects *[or all of them?]* that person is in.
> - Next, add that person to your 'roster' (the list of people available to chat to)
> - Finally, click their name in the chat interface and start a chat!

Note: I haven't been able to test the chat feature yet, so I've not really an idea about the actual steps. I'm just proposing a more general question with a series of answers for the individual steps required to start chatting with someone; you'll have to write text which is accurate to the actual use case, if what I've said above isn't accurate.

If chat is not automatically activated for each user when a podmin enables it in the pod, the first step might be:

> - If you haven't already enabled the chat feature, go to your user settings page to do this.

Does this sound like a useful approach?

Flaburgan replied:
No need to add the contact to the roster, once you activated the chat for an aspect, every contacts in this aspect are immediately available to chat with.

AugierLe42e replied:
> Finally, click their name in the chat interface and start a chat!

I think they'll that point by themselves ^_^

So, I'll propose :

> ### How do I chat with someone in diaspora*?
You need to enable chat for one of the aspects that person is in. To do so, go to the contacts page, select an aspect and click on the chat icon to enable chat for the aspect. Or, you can create a new one. Then you can choose to enable chat for it by enabling the option [here, get the text displayed in the Facebox]

What do you think about it ?

goobertron replied:
We're nearly there! How about:

> ### How do I chat with someone in diaspora*?
> 
> First, you need to enable chat for one of the aspects that person is in. To do so, go to the %{contacts_page}, select the aspect you want and click on the chat icon to enable chat for the aspect. You could, if you prefer, create a special aspect called 'Chat' and add the people you want to chat with to that aspect. Once you've done this, open the chat interface and select the person you want to chat with.

I don't think any of the rest of the in-app help section contains images, so I wouldn't recommend adding them here. However, it would also be good to add a section about chat to the tutorials on the project site. To do that, make a PR in [github](https://github.com/diaspora/diaspora-project-site/tree/master/app/views/pages). I'll work with you on that for the English, if you like.

AugierLe42e replied:
It's not images it's just entypo icons ;)
Ok to me, I update the text.

goobertron replied:
You forgot to update the question as well as the answer...

AugierLe42e commented:
@Zauberstuhl : whil working on this, maybe you want to add something ? Like how to connect with an external client ?

Zauberstuhl commented:
we can add items if it is merged. I have no access to your branch :)

goobertron commented on config/locales/diaspora/en.yml:
```diff
@@ -449,6 +449,11 @@ en:
+    chat:
+      title: "Chat"
+      contacts_page: "contacts page"
+      add_contact_roster_q: "How do I add contacts to my roster?"
```
Should be "How do I chat with someone in diaspora*?"

AugierLe42e commented on app/controllers/help_controller.rb:
```diff
@@ -1,4 +1,8 @@
+  def faq
+    gon.chatEnabled = AppConfig.chat.enabled?
+  end
+end
```
You are not very kind to me, Milou... Always criticizing me :',(

AugierLe42e commented:
Looks ok to me now.

jhass commented:
Please squash commits. Tipp: You can edit a commit with `git commit --amend`

jhass commented on app/assets/javascripts/app/views/help_view.js:
```diff
@@ -35,6 +36,13 @@ app.views.Help = app.views.StaticContentView.extend({
+    this.CHAT_SUBS = {
+      add_contact_roster_a: {
+        toggle_privilege: this.getChatIcons(),
+        contacts_page: this.linkHtml('/contacts', Diaspora.I18n.t('chat.contacts_page'))
```
Oh, actually we wanted to start using `js-routes` for these.

jhass commented:
Thanks!
